File: pre/c基础/code/GifDrawable.java
package com.zero.gifframe;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import android.os.SystemClock;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;


public class GifDrawable extends Drawable implements Animatable, Runnable {

    private final GifFrame gifFrame;
    private final BitmapProvider bitmapProvider;
    private final Rect srcRect;
    private Bitmap bitmap;
    private final Paint paint;
    private final int frameCount;
    private int frameIndex;
    private boolean isRunning;
    long showtime, curTime;

    public interface BitmapProvider {
        Bitmap acquireBitmap(int minWidth, int minHeight);

        void releaseBitmap(Bitmap bitmap);
    }

    public GifDrawable(GifFrame gifFrame, BitmapProvider bitmapProvider) {
        this.gifFrame = gifFrame;
        this.bitmapProvider = bitmapProvider;

        int width = gifFrame.getWidth();
        int height = gifFrame.getHeight();
        frameCount = gifFrame.getFrameCount();
        frameIndex = 0;

        bitmap = bitmapProvider.acquireBitmap(width, height);
        paint = new Paint();
        paint.setFilterBitmap(true);
        gifFrame.getFrame(bitmap, getFrameIndex());
        srcRect = new Rect(0, 0, width, height);
    }

    @Override
    public void start() {
        if (!isRunning()) {
            isRunning = true;
            scheduleSelf(this, SystemClock.uptimeMillis());
        }
    }

    @Override
    public void stop() {
        if (isRunning()) {
            isRunning = false;
            unscheduleSelf(this);
        }
    }

    @Override
    public boolean isRunning() {
        return isRunning;
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        canvas.drawBitmap(bitmap, srcRect, getBounds(), paint);
        showtime = SystemClock.uptimeMillis();
        if (isRunning()) {
            scheduleSelf(this, showtime + curTime);
        }
    }

    @Override
    public void setAlpha(int i) {
        paint.setAlpha(i);
    }

    @Override
    public void setColorFilter(@Nullable ColorFilter colorFilter) {
        paint.setColorFilter(colorFilter);
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSPARENT;
    }

    @Override
    public void run() {
        curTime = gifFrame.getFrame(bitmap, getFrameIndex());
        invalidateSelf();
    }

    public int getFrameIndex() {
        frameIndex++;
        return frameIndex < frameCount ? frameIndex : (frameIndex = 0);
    }
}

File: source/Giflib1/app/src/main/java/com/zero/giflib1/gif/GifDrawable.kt
package com.zero.giflib1.gif

import android.graphics.*
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.os.SystemClock


class GifDrawable(private val mGifFrame: GifFrame) : Drawable(),
    Animatable, Runnable {
    private val mPaint: Paint
    //1. 首先Java准备一个Bitmap传给C++
    //2. gif 解压出来之后每一帧(Screen) 填充 成Bitmap
    //3. 成Bitmap 绘制到Canvas
    //4. Bitmap是要Java端初始化出来的
    private val mBitmap: Bitmap //成员属性 Java成员变量 + get + set
    private val srcRect //gif的矩形区域
            : Rect
    //gif在C++解码  width,height
    val width: Int = mGifFrame.width
    val height: Int = mGifFrame.height
    private var runFlag = false
    private var showTime: Long = 0
    private var curTime //当前帧要显示的时间 C++ 获取的
            : Long = 0
    private var frameCount: Int = mGifFrame.frameCount
    private var frameIndex: Int
        get() {
            field++
            return if (field < frameCount) field else 0.also { field = it }
        }


    init {
        mBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        mPaint = Paint()
        mPaint.isFilterBitmap = true
        frameIndex = 0
        srcRect = Rect(0, 0, width, height)
        mGifFrame.getFrame(mBitmap, frameIndex)
    }


    override fun draw(canvas: Canvas) { //绘制 把gif的每一帧绘制到canvas
        canvas.drawBitmap(mBitmap, srcRect, bounds, mPaint)
        showTime = SystemClock.uptimeMillis()
        if (isRunning) {
            scheduleSelf(this, showTime + curTime)
        }
    }

    override fun setAlpha(i: Int) {
        mPaint.alpha = i
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        mPaint.colorFilter = colorFilter
    }

    override fun getOpacity(): Int {
        return PixelFormat.TRANSLUCENT
    }

    override fun start() {

        if (!isRunning) {
            runFlag = true
            scheduleSelf(this, SystemClock.uptimeMillis())
        }
    }

    override fun stop() {
        if (isRunning) {
            runFlag = false
            unscheduleSelf(this)
        }
    }

    override fun isRunning(): Boolean {
        return runFlag
    }

    override fun run() { //获取每一帧
        curTime = mGifFrame.getFrame(mBitmap, frameIndex)
        invalidateSelf()
    }

}